test/integration/garbage_collector_test.go

```diff
@@ -0,0 +1,159 @@
+			Name:      podName,
+			Namespace: framework.TestNS,
+			OwnerReferences: []v1.OwnerReference{
+				{
+					Kind:       "ReplicationController",
+					Name:       "non-exist-owner",
+					UID:        "123",
+					APIVersion: "v1",
+				},
+			},
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:  "fake-name",
+					Image: "fakeimage",
+				},
+			},
+		},
+	}
+}
+
+func newOwnerRC() *v1.ReplicationController {
+	return &v1.ReplicationController{}
+}
+
+func observePodDeletion(t *testing.T, w watch.Interface) (deletedPod *api.Pod) {
+	deleted := false
+	timeout := false
+	timer := time.After(60 * time.Second)
+	for !deleted && !timeout {
+		select {
+		case event, _ := <-w.ResultChan():
+			if event.Type == watch.Deleted {
+				// TODO: used the commented code once we fix the client.
+				// deletedPod = event.Object.(*v1.Pod)
+				deletedPod = event.Object.(*api.Pod)
+				deleted = true
+			}
+		case <-timer:
+			timeout = true
+		}
+	}
+	if !deleted {
+		t.Fatalf("Failed to observe pod deletion")
+	}
+	return
+}
+
+func TestCascadingDeletion(t *testing.T) {
+	flag.Set("v", "9")
+	var m *master.Master
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		m.Handler.ServeHTTP(w, req)
+	}))
+	// TODO: Uncomment when fix #19254
+	// defer s.Close()
+
+	masterConfig := framework.NewIntegrationTestMasterConfig()
+	m, err := master.New(masterConfig)
+	if err != nil {
+		t.Fatalf("Error in bringing up the master: %v", err)
+	}
+
+	framework.DeleteAllEtcdKeys()
+	clientSet, err := clientset.NewForConfig(&restclient.Config{Host: s.URL})
+	if err != nil {
+		t.Fatalf("Error in create clientset: %v", err)
+	}
+	groupVersionResources, err := garbagecollector.ServerPreferredGroupVersionResources(clientSet.Discovery())
+	if err != nil {
+		t.Fatalf("Failed to get supported resources from server: %v", err)
+	}
+	clientPool := dynamic.NewClientPool(&restclient.Config{Host: s.URL}, dynamic.LegacyAPIPathResolverFunc)
+	gc, err := garbagecollector.NewGarbageCollector(clientPool, groupVersionResources)
+	if err != nil {
+		t.Fatalf("Failed to create garbage collector")
+	}
+
+	podClient := clientSet.Core().Pods(framework.TestNS)
+	_, err = podClient.Create(newDanglingPod())
+	if err != nil {
+		t.Fatalf("Failed to create Pod: %v", err)
+	}
+
+	// set up watch
+	pods, err := podClient.List(api.ListOptions{})
+	if err != nil {
+		t.Fatalf("Failed to list pods: %v", err)
+	}
+	if len(pods.Items) != 1 {
+		t.Fatalf("Expect only 1 pod")
+	}
+	options := api.ListOptions{
+		ResourceVersion: pods.ListMeta.ResourceVersion,
+	}
+	w, err := podClient.Watch(options)
+	if err != nil {
+		t.Fatalf("Failed to set up watch: %v", err)
+	}
+	stopCh := make(chan struct{})
+	go gc.Run(5, stopCh)
+	defer close(stopCh)
+	deletedPod := observePodDeletion(t, w)
+	if deletedPod == nil {
+		t.Fatalf("empty deletedPod")
+	}
+	if deletedPod.Name != podName {
+		t.Fatalf("deleted unexpected pod: %v", *deletedPod)
+	}
```

Comment:
Add all varieties of pods which you *don't* want the GC to delete and verify that it doesn't delete them.